cql3/statements/ks_prop_defs.cc

```diff
@@ -226,7 +226,8 @@ lw_shared_ptr<data_dictionary::keyspace_metadata> ks_prop_defs::as_ks_metadata_u
         options = old_options;
     }
     std::optional<unsigned> initial_tablets = get_initial_tablets(*sc, old->initial_tablets().has_value());
-    if (!initial_tablets) {
+    // if tablets options have not been specified in ALTER KS, inherit them if it's tablets-enabled KS
+    if (initial_tablets == 0 && old->initial_tablets().has_value()) {
```

Comment:
There should be 3 states returned from get_initial_tablets()

1. user asks for tablets ON -- this is `initial_tablets.has_value()`
2. user doesn't asks any specific -- this is `initial_tablets.has_value() && initial_tablets.value() == 0`
3. user asks for tablets OFF -- this is `!initial_tablets.has_value()`

And only case 2. calls for inheritance

You check for `initial_tablets == 0` which is `initial_tablets.has_value() && initial_tablets.value() == 0` which partially overlaps with case 1 -- e.g. when `TABLETS { enabled = true }` or `TABLETS { initial = 0 }` is specified

Reply:
OK, i believe it should be fixed now.